packages/grid/x-data-grid/src/internals/hooks/features/rows/useGridRows.ts

```diff
@@ -97,33 +76,63 @@ const getRowsStateFromCache = (
   rowsCache: GridRowsInternalCache,
   previousTree: GridRowTreeConfig | null,
   apiRef: React.MutableRefObject<GridApiCommunity>,
+  rowCountProp: number | undefined,
 ): GridRowsState => {
-  const {
-    props: { rowCount: propRowCount = 0 },
-    value,
-  } = rowsCache.state;
+  const { value } = rowsCache.state;
+  const rowCount = rowCountProp ?? 0;
 
   const groupingResponse = apiRef.current.unstable_groupRows({ ...value, previousTree });
 
   const dataTopLevelRowCount = Object.values(groupingResponse.tree).filter(
     (node) => node.parent == null,
   ).length;
   const totalRowCount =
-    propRowCount > groupingResponse.ids.length ? propRowCount : groupingResponse.ids.length;
-  const totalTopLevelRowCount =
-    propRowCount > dataTopLevelRowCount ? propRowCount : dataTopLevelRowCount;
+    rowCount > groupingResponse.ids.length ? rowCount : groupingResponse.ids.length;
+  const totalTopLevelRowCount = rowCount > dataTopLevelRowCount ? rowCount : dataTopLevelRowCount;
 
   return { ...groupingResponse, totalRowCount, totalTopLevelRowCount };
 };
 
+export const rowsStateInitializer: GridStateInitializer<
+  Pick<DataGridProcessedProps, 'rows' | 'rowCount' | 'getRowId'>
+> = (state, props, apiRef) => {
+  const rowsCache = {
+    state: convertGridRowsPropToState({
+      rows: props.rows,
+      getRowId: props.getRowId,
+      prevState: {
+        value: {
+          idRowsLookup: {},
+          ids: [],
+        },
+        rowsBeforePartialUpdates: [],
+      },
+    }),
+    timeout: null,
+    lastUpdateMs: Date.now(),
+  };
+
+  return {
+    ...state,
+    rows: getRowsStateFromCache(rowsCache, null, apiRef, props.rowCount),
+    rowsCache,
```

Comment:
Can we add a `TODO` somewhere to say that we want to remove `rowsCache` from the state ?
I am in favor of doing it in a follow up PR with a clean structure.